var/spack/repos/builtin/packages/plumed/package.py

```diff
@@ -73,58 +74,41 @@ class Plumed(AutotoolsPackage):
     depends_on('automake', type='build')
     depends_on('libtool', type='build')
 
-    # Dictionary mapping PLUMED versions to the patches it provides
-    # interactively
-    plumed_patches = {
-        '2.3.0': {
-            'amber-14': '1',
-            'gromacs-2016.1': '2',
-            'gromacs-4.5.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.4': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        },
-        '2.2.4': {
-            'amber-14': '1',
-            'gromacs-4.5.7': '2',
-            'gromacs-4.6.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.2': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        },
-        '2.2.3': {
-            'amber-14': '1',
-            'gromacs-4.5.7': '2',
-            'gromacs-4.6.7': '3',
-            'gromacs-5.0.7': '4',
-            'gromacs-5.1.2': '5',
-            'lammps-6Apr13': '6',
-            'namd-2.8': '7',
-            'namd-2.9': '8',
-            'espresso-5.0.2': '9'
-        }
-    }
-
     force_autoreconf = True
 
     parallel = False
 
     def apply_patch(self, other):
-        plumed = subprocess.Popen(
-            [join_path(self.spec.prefix.bin, 'plumed'), 'patch', '-p'],
-            stdin=subprocess.PIPE
+
+        # The name of MD engines differ slightly from the ones used in Spack
+        format_strings = collections.defaultdict(
+            lambda: '{0.name}-{0.version}'
+        )
+        format_strings['espresso'] = 'q{0.name}-{0.version}'
+        format_strings['amber'] = '{0.name}{0.version}'
+
+        get_md = lambda x: format_strings[x.name].format(x)
+
+        # Get available patches
+        plumed_patch = Executable(
+            join_path(self.spec.prefix.bin, 'plumed-patch')
         )
-        opts = Plumed.plumed_patches[str(self.version)]
-        search = '{0.name}-{0.version}'.format(other)
-        choice = opts[search] + '\n'
-        plumed.stdin.write(choice)
-        plumed.wait()
+
+        out = plumed_patch('-q', '-l', output=str)
+        available = out.split(':')[-1].split()
+
+        # Check that `other` is among the patchable applications
+        if get_md(other) not in available:
+            msg = '{0.name}@{0.version} is not among the MD engine'
+            msg += ' that can be patched by {1.name}@{1.version}.\n'
+            msg += 'Supported engines are:\n'
+            for x in available:
+                msg += x + '\n'
+            raise RuntimeError(msg.format(other, self.spec))
+
+        # Call plumed-patch to patch executables
+        target = format_strings[other.name].format(other)
+        plumed_patch('-p', '-e', target)
```

Comment:
Does this command patch the source or the executable?

Reply:
The sources...

Reply:
In `gromacs`, for instance, I added the following:
```python
def patch(self):
    if '+plumed' in self.spec:
        self.spec['plumed'].package.apply_patch(self)
```
which patches `gromacs` sources with a patch shipped with plumed (a dependency). Not all codes that use `plumed` need to be patched, but [some do according to docs](https://plumed.github.io/doc-v2.3/user-doc/html/index.html).

Reply:
Ok, I see, thanks!

I was just wondering if you could do something like:
`depends_on("gromacs", patches=patch('https://github.com/plumed/plumed2/blob/master/patches/gromacs-4.5.7.diff', when='@4.5.7', level=1))`

Reply:
Variant + optional dependency makes sense for this case.